### ISSUE TITLE
[Feature Request] - Bazel Caching validity and management

### ISSUE THREAD

**Issue by DEKHTIARJonathan** [bazelbuild/bazel#11182]
I have a few questions which might be feature requests depending if it's doable or not:

#### Question 1: Is there a way to control the maximum file size being remotely uploaded & cached ?

The final binary of some large projects might be very large (hundreds of MBs)

Do we have a way to say: I don't want to cache any file larger than 15MBs for instance ? 
This is especially important/useful when the network bandwidth is not very strong. It would allow to benefit from downloading any cache available, but avoid pushing very large files back to the cache.

#### Question 2: Is there a way to force overwrite cache if existing ?

In some cases we might have corrupted caches, how could we tell bazel: please do not care about existing cache and overwrite existing caches (if they exists).

#### Question 3: How could we push local caches to remote if they don't exist ?

In some cases the connection to the remote cache server might drop or the cache server itself might down for a few minutes. How could we say Bazel to resync local and remote cache or rebuild with remote cache only and push whatever is missing if they are available in local cache.

Thanks for the help

**Comment by ulfjack** [comment]
Question 1: Not that I'm aware. Should be easy to add if we want it.

Question 2: run bazel clean and then bazel build --no_remote_accept_cached (I'm not 100% sure if that applies to all remote cache implementations)

Question 3: What for? It doesn't seem like doing that will be a significant win, unless you expect the connection / server to be unavailable for a significant fraction of the time, in which case you probably have bigger problems?

**Comment by ulfjack** [comment]
I think for Q1, you could just configure the cache to refuse large files rather than configuring every single client independently. That might be a better solution than adding another feature to Bazel.

**Comment by DEKHTIARJonathan** [comment]
Thanks @ulfjack for the answer.

- Q1: Your proposal to enforce this server-side is not always desirable. We may want to only restrict the upload of large caches to specific workers (those with slow internet connection).

I'll give you an example. Currently with COVID-19, many engineers WFH. We would like to keep full caching for the servers / workstations building on-site, and limit upload for those building from home. Uploading 100MB+ files from a standard home connection is not really ideal.

- Q2: Just to make sure I understand what this flag do. If you activate `--no_remote_accept_cached` you would essentially put the cache in "Write-only" mode (no read). Correct ?

- Q3: Servers, VPN, network might be down for some of the build time or even the totality of one/many builds. Thus bazel would not cache. Is there away to ask bazel to upload the local caches remotely ? That could be very useful in a distributed build environment (not necessarily in a datacenter environment).

Thanks a lot for your answers ;) Much appreciated

**Comment by ulfjack** [comment]
- Q1: central administration doesn't prevent you from applying different policies to different clients. You could either go by IP or use some form of authentication and grouping.

- Q2: I believe so.

- Q3: There isn't right now, no. I'm not convinced it's useful, though. If there's a missing cache entry, then the next build will just upload the entry, and then you're back to normal. Uploading the local caches just doesn't seem like a significant win.

**Comment by alexeagle** [comment]
I have the same problem as Q1.

> I think for Q1, you could just configure the cache to refuse large files rather than configuring every single client independently. That might be a better solution than adding another feature to Bazel.

Interesting!

I proposed to make the server reject such uploads in https://github.com/buchgr/bazel-remote/issues/440 and the conclusion over there was that due to the protocol, the server doesn't have enough info to know what action is responsible for the large cache entry. That means that we would be missing a feedback loop to find and fix actions that produce giant files (in our case it was misuse of rules_docker resulting in entire OCI images being action inputs, rather than just layers)

It's certainly possible to give up on identifying these bad actions though, and just leave the rejected upload / warning behavior.

The remaining problem is the network traffic caused by uploading the bytes which are then dropped on the floor (though I'll have to study the protocol to understand if the blob size is reported early enough that the server could hang up before the upload happens).

I wonder if those maintaining the client would rather have features added to the server, and those maintaining the server would rather have features added to the client?

**Comment by alexeagle** [comment]
Let's see if it really sends the data to the server on those 6 attempts to upload. I installed wireshark, and am using it for the first time so forgive my clumsiness.
I patched bazel-remote like this https://gist.github.com/alexeagle/71d39f470ba45c7f8a7fcc38a70bfd8c to reject uploads over 10MB

Simple test fixture is like
```
genrule(
    name = "large-output",
    srcs = [],
    outs = ["out"],
    cmd = "head -70000 /dev/random > $@",
)
```

I recorded one session where bazel uploads a ~12M byte artifact to a locally running bazel-remote, and wireshark shows cumulative bytes sent over loopback at 996k bytes. So that's good, looks like it terminates the upload partway through the bytestream. Just to be sure, I also did an experiment uploading a  ~18M byte artifact, and see cumulative bytes to remote-cache around 420k.

So if we could just hide the BulkTransferException from scaring the user, I think this method works.

/cc @mostynb for any thoughts or point out something I did wrong.
